Stop repair scheduler if two major versions are detected

himanshujindal opened the pull request:
Do not allow repair scheduler to run repair if two major versions are detected. This can happen during major version upgrade, and we don't want repair to run since the streaming data for repair can be incompatible across major versions.

Users can override this using a cassandra yaml config if they want to.

Also, this does not affect minor version upgrades, or repair jobs run manually.

https://issues.apache.org/jira/browse/CASSANDRA-20048

Thanks for sending a pull request! Here are some tips if you're new here:
 
 * Ensure you have added or run the [appropriate tests](https://cassandra.apache.org/_/development/testing.html) for your PR.
 * Be sure to keep the PR description updated to reflect all changes.
 * Write your PR title to summarize what this PR proposes.
 * If possible, provide a concise example to reproduce the issue for a faster review.
 * Read our [contributor guidelines](https://cassandra.apache.org/_/development/index.html)
 * If you're making a documentation change, see our [guide to documentation contribution](https://cassandra.apache.org/_/development/documentation.html)
 
Commit messages should follow the following format:

```
<One sentence description, usually Jira title or CHANGES.txt summary>

<Optional lengthier description (context on patch)>

patch by <Authors>; reviewed by <Reviewers> for CASSANDRA-#####

Co-authored-by: Name1 <email1>
Co-authored-by: Name2 <email2>

```

The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

himanshujindal committed:
Stop repair scheduler if two major versions are detected

Do not allow repair scheduler to run repair if two major versions are detected.
This can happen during major version upgrade, and we don't want repair to run
since the streaming data for repair can be incompatible across major versions.

Users can override this using a cassandra yaml config if they want to.

Also, this does not affect minor version upgrades, or repair jobs run manually.

https://issues.apache.org/jira/browse/CASSANDRA-20048

himanshujindal commented:
This is based off of the existing PR from clohfink@ - https://github.com/jaydeepkumar1984/cassandra/pull/20/commits/c63a5a4d9881417f89d5e9836dbb8045dcdb558e

Changes from that PR
* Used yaml config instead of guardrail
* Stop auto repair only, not the manual repair jobs
* Stop repair for multiple major versions only, and not minor versions

jaydeepkumar1984 commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -374,6 +374,9 @@ public MemtableOptions()
     public volatile double incremental_repair_disk_headroom_reject_ratio = 0.2;
 
+    // by default repair is disabled if there are mixed major versions detected, but you can enable it using this flag
+    public volatile boolean mixed_major_version_repair_enabled = false;
```
Can you add the configuration inside AutoRepairConfig.java?

himanshujindal replied:
Yes. Will address in next revision.

jaydeepkumar1984 commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:
```diff
@@ -165,6 +166,11 @@ public void repair(AutoRepairConfig.RepairType repairType)
         }
+        if (!DatabaseDescriptor.getMixedMajorVersionRepairEnabled() &&
+            Gossiper.instance.hasMultipleLiveMajorVersions()) {
+            logger.info("Auto-repair is disabled when nodes in the cluster have different major versions");
```
Please add a metric

himanshujindal replied:
I am not seeing this class emit metrics. I'll be happy to introduce metrics to this class. Do you have an example of code where we do emit metric, so I can use that pattern to emit metrics?

jaydeepkumar1984 replied:
https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/metrics/AutoRepairMetrics.java

tolbertam replied:
I think the metrics are calculated off of / engaged by  `AutoRepairState` (accessed 2 lines below).  You could add something to that.   Curious on the metric we are interested in adding?  I'm guess maybe a `Counter` that increments whenever repair is skipped because mixed versions are detected?

jaydeepkumar1984 replied:
The metric is still missing?

himanshujindal replied:
@jaydeepkumar1984 See my comment here - https://github.com/apache/cassandra/pull/4304#issuecomment-3226130821 and Andy's reply in slack - https://the-asf.slack.com/archives/C06MR8K2Q0Z/p1756306978752409?thread_ts=1755038748.234549&cid=C06MR8K2Q0Z

We decided not to add a metric since logging should be sufficient here. Having said that, I will be open to adding a metric if you align with my proposal presented in the comment linked.

jaydeepkumar1984 replied:
Sounds good!

jaydeepkumar1984 commented on src/java/org/apache/cassandra/gms/Gossiper.java:
```diff
@@ -2034,6 +2034,23 @@ private boolean nodesAgreeOnSchema(Collection<InetAddressAndPort> nodes)
         return true;
     }
 
+    public boolean hasMultipleLiveMajorVersions()
```
Can we extend the existing `AutoRepairUtils.java::getHostIdsInCurrentRing`, which already goes through all the nodes in a cluster?

https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java#L429

himanshujindal replied:
Yes, added in new revision.

jaydeepkumar1984 commented on test/distributed/org/apache/cassandra/distributed/test/GossipTest.java:
```diff
@@ -285,4 +288,40 @@ public void testQuarantine() throws IOException
+                               .start())
+        {
+            InetSocketAddress peer1 = cluster.get(1).broadcastAddress();
+            String version1 = "4.0.1920";
```
Can you add a test case by setting different minor versions explicitly, say `4.0.1920`, `4.1.2`?

himanshujindal replied:
Yes, added in new revision.

himanshujindal committed:
Address feedback from Jaydeep

himanshujindal commented:
@jaydeepkumar1984 I have addressed all the feedback except metrics, which I can do in the next revision if you can point me to a code link.

For adding tests, I found this class called ClusterMetadataTestHelper, which I was able to use and extend. This worked really well. Let me know if there is any feedback with how I wrote those tests.

himanshujindal committed:
Fix lint issues

himanshujindal committed:
Fix checkstyle

tolbertam commented on conf/cassandra.yaml:
```diff
@@ -2779,6 +2779,9 @@ storage_compatibility_mode: NONE
 #   # NOTE: Each of the below settings can be overridden per repair type under repair_type_overrides
+#   # by default repair is disabled if there are mixed major versions detected - which would happen
+#   # if a major version upgrade is being performed on the cluster, but a user can enable it using this flag
+#   # mixed_major_version_repair_enabled = false;
```
Small suggestions:

* Move `#NOTE: ...` to the line before `global_settings` as it makes more sense to be proximate to there.
* Uncomment `mixed_major_version_repair_enabled = false` and remove semi-colon.

```suggestion
#   # by default repair is disabled if there are mixed major versions detected - which would happen
#   # if a major version upgrade is being performed on the cluster, but a user can enable it using this flag
#   mixed_major_version_repair_enabled = false
#   # NOTE: Each of the below settings can be overridden per repair type under repair_type_overrides
```

A couple of other things that would to be good to do:

1. Can you also add this to `cassandra_latest.yaml` as well?
2. Add to 'Top level settings' in auto_repair.adoc? In that file it would also be good to expand on why repairs should not be run during major version upgrades (previous slack thread about it: https://the-asf.slack.com/archives/C06MR8K2Q0Z/p1729716524563759)

himanshujindal replied:
Fixed.

tolbertam commented on src/java/org/apache/cassandra/config/DatabaseDescriptor.java:
```diff
@@ -5940,6 +5940,16 @@ public static void setIncrementalRepairDiskHeadroomRejectRatio(double value)
+    {
+        conf.auto_repair.mixed_major_version_repair_enabled = mixed_major_version_repair_enabled;
+    }
+
```
It looks like we already have a `AutoRepairConfig.isMixedMajorVersionRepairEnabled()` which is currently unused.   Since this is an `AutoRepair`-specific thing can we remove these methods and use methods on `AutoRepairConfig` instead?

Would also be nice to update `SetAutoRepairConfig` to allow configuring this value.

himanshujindal replied:
Thanks for pointing me to to that tool. Updated that tool to read this config as well.

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java:
```diff
@@ -425,6 +425,21 @@ public static CurrentRepairStatus getCurrentRepairStatus(RepairType repairType,
+     *  false if only one major version is detected
+     *
+     */
+    public static boolean hasMultipleLiveMajorVersions()
+    {
+        ClusterMetadata metadata = ClusterMetadata.current();
+        int maxMajorVersion = ClusterMetadata.current().directory.clusterMaxVersion.cassandraVersion.major;
+        int minMajorVersion = ClusterMetadata.current().directory.clusterMinVersion.cassandraVersion.major;
+        return maxMajorVersion != minMajorVersion;
+    }
+
```
👍 nice and simple.

I had some doubts on whether this is sufficient given Cassandra's version history.   e.g. between 2.0 and 2.1 maybe this wouldn't be safe?  But I'd hope from C* 6.0 on this is probably sufficient.

himanshujindal replied:
Yes, I think so too. My understanding is that we only want to block repair on multiple major versions right now.

tolbertam commented on test/distributed/org/apache/cassandra/distributed/test/log/ClusterMetadataTestHelper.java:
```diff
@@ -798,19 +798,27 @@ public MoveProcess finishMove()
+        addEndpoint(i, NodeVersion.CURRENT);
+    }
+
+    public static void addEndpoint(int i, NodeVersion nodeVersion) {
         try
```
`{` on next line

```suggestion
    public static void addEndpoint(int i, NodeVersion nodeVersion)
    {
        try
```

himanshujindal replied:
Fixed.

tolbertam commented on test/distributed/org/apache/cassandra/distributed/test/log/ClusterMetadataTestHelper.java:
```diff
@@ -798,19 +798,27 @@ public MoveProcess finishMove()
     }
 
-    public static void addEndpoint(InetAddressAndPort endpoint, Token t)
+    public static void addEndpoint(InetAddressAndPort endpoint, Token t) {
+        addEndpoint(endpoint, t, NodeVersion.CURRENT);
```
{ on next line:

```suggestion
    public static void addEndpoint(InetAddressAndPort endpoint, Token t)
    {
        addEndpoint(endpoint, t, NodeVersion.CURRENT);
```

himanshujindal replied:
Fixed.

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairUtilsTest.java:
```diff
@@ -232,6 +236,52 @@ public void testGetHostIdsInCurrentRing_multiple_nodes()
+    @Test
+    public void testHasMultipleLiveMajorVersionsWithMultipleNodesOfDifferentMajorVersions()
+    {
+        // add two nodes with cassandra 5 major version, but different minor version
```
A bit pedantic I know, but in future the current major version will change.

```suggestion
        // add two nodes with current cassandra major version, but different minor version
```

himanshujindal replied:
Fixed.

tolbertam commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairUtilsTest.java:
```diff
@@ -232,6 +236,52 @@ public void testGetHostIdsInCurrentRing_multiple_nodes()
+    @Test
+    public void testHasMultipleLiveMajorVersionsWithMultipleNodesOfSameMajorVersionDifferentMinorVersions()
+    {
+        // add two nodes with cassandra 5 major version, but different minor version
```
```suggestion
        // add two nodes with current cassandra major version, but different minor version
```

himanshujindal replied:
Fixed.

tolbertam reviewed (commented):
Change looks good, nice and straightforward!   Had a few suggestions and happy to follow up with another review based on further changes.

himanshujindal committed:
Address tolbertam@ feedback

himanshujindal committed:
Address build issues

himanshujindal committed:
Address build issues

himanshujindal committed:
Fix comments in cassandra_latest.yaml

himanshujindal committed:
Address more tolbertam@ feedback

himanshujindal commented:
@tolbertam Thanks for your feedback. I have addressed all of it except adding metrics. Let me know if that is a blocker and I can get to it next week.

If metrics are a concern, does this plan sound good?
I’m planning to add a new metric called repairSkippedDueToMixedMajorVersions. It will track when repairs are skipped during Cassandra cluster upgrades, and when they resume. While users could look at overall repair metrics, this specific metric makes it clear that repairs were skipped because of mixed major versions, not for other reasons..

Also, I have run pre-ci testing on the commit and the tests look good - https://pre-ci.cassandra.apache.org/job/cassandra/74/testReport/

himanshujindal committed:
Merge branch 'trunk' into CASSANDRA-20048

tolbertam commented on doc/modules/cassandra/pages/managing/operating/auto_repair.adoc:
```diff
@@ -219,6 +219,10 @@ in nodetool repair. General advice is to keep this true.
 | repair_retry_backoff | 30s | Backoff time before retrying a repair session.
 | repair_task_min_duration | 5s | Minimum duration for the execution of a single repair task. This prevents the
 scheduler from overwhelming the node by scheduling too many repair tasks in a short period of time.
+| mixed_major_version_repair_enabled | false | Enable/Disable running repairs on the cluster when there are mixed.
```
Should remove extraneous `.` here.

```suggestion
| mixed_major_version_repair_enabled | false | Enable/Disable running repairs on the cluster when there are mixed
```

himanshujindal committed:
Update doc/modules/cassandra/pages/managing/operating/auto_repair.adoc

Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>

tolbertam commented on doc/modules/cassandra/pages/managing/operating/auto_repair.adoc:
```diff
@@ -219,6 +219,10 @@ in nodetool repair. General advice is to keep this true.
 | repair_task_min_duration | 5s | Minimum duration for the execution of a single repair task. This prevents the
 scheduler from overwhelming the node by scheduling too many repair tasks in a short period of time.
+| mixed_major_version_repair_enabled | false | Enable/Disable running repairs on the cluster when there are mixed.
+major versions detected, which usually occurs when the cluster is being upgraded. Repair is not supported in this scenario
```
```
Repair is not supported in this scenario as data might be streamed from a node with a different version of Cassandra leading to compatibility issues. Once major versions are compatible for streaming data, this setting can be turned to true.
```

Could we just rephrase this as:

```
Repairs between nodes of different major versions is not something that is tested, so this may lead to data compatibility issues.  It is strongly discouraged to set this to true without doing extensive testing beforehand.
```

himanshujindal committed:
Fix documentation

himanshujindal commented:
Here's the test run using the latest commit - https://pre-ci.cassandra.apache.org/job/cassandra/lastBuild/testReport/

tolbertam commented on src/java/org/apache/cassandra/tools/nodetool/SetAutoRepairConfig.java:
```diff
@@ -53,7 +53,7 @@ public class SetAutoRepairConfig extends AbstractCommand
                                                           "|materialized_view_repair_enabled|repair_max_retries" +
                                                           "|repair_retry_backoff|repair_session_timeout|min_repair_task_duration" +
-                                                          "|repair_by_keyspace|token_range_splitter.<property>]" })
+                                                          "|repair_by_keyspace|token_range_splitter.<property>]|mixed_major_version_repair_enabled" })
```
```suggestion
                                                          "|repair_by_keyspace|mixed_major_version_repair_enabled|token_range_splitter.<property>]" })
```

Suggest moving it before `token_range_splitter.<property>` since that's a bit of an exceptional thing.  Really we should order this alphabetically, but I think its fine as is for now.

himanshujindal committed:
Update src/java/org/apache/cassandra/tools/nodetool/SetAutoRepairConfig.java

Co-authored-by: Andrew Tolbert <6889771+tolbertam@users.noreply.github.com>

tolbertam commented on src/java/org/apache/cassandra/tools/nodetool/SetAutoRepairConfig.java:
```diff
@@ -176,6 +176,9 @@ public void execute(NodeProbe probe)
             case "repair_retry_backoff":
                 probe.setAutoRepairRetryBackoff(repairTypeStr, paramVal);
                 break;
+            case "mixed_major_version_repair_enabled":
```
Can we add the output to `getautorepairconfig` as well? Looks like setting this works, but would be nice to verify that it works.

tolbertam replied:
Ah, also like the documentation, since this is not repairType driven, it should be moved to the switch statement starting at line 86 so we aren't required to provide a repairType to set this.

himanshujindal replied:
Good catch. These should be top level settings and not auto repair settings. Fixed.

tolbertam commented on doc/modules/cassandra/pages/managing/operating/auto_repair.adoc:
```diff
@@ -219,6 +219,10 @@ in nodetool repair. General advice is to keep this true.
+| mixed_major_version_repair_enabled | false | Enable/Disable running repairs on the cluster when there are mixed
+major versions detected, which usually occurs when the cluster is being upgraded. Repairs between nodes of
+different major versions is not something that is tested, so this may lead to data compatibility issues.
+It is strongly discouraged to set this to true without doing extensive testing beforehand.
```
Just realized, since this is not a repair-type level setting, it should be moved up to 'Top level settings' right?

himanshujindal replied:
Good catch. These should be top level settings and not auto repair settings. Fixed.

himanshujindal committed:
Move config key to top level settings in doc and tools

tolbertam commented on test/unit/org/apache/cassandra/tools/nodetool/SetAutoRepairConfigTest.java:
```diff
@@ -268,7 +268,8 @@ public static Collection<Object[]> testCases()
             forEachRepairType("repair_max_retries", "3", (type) -> verify(probe, times(1)).setAutoRepairMaxRetriesCount(type.name(), 3)),
-            forEachRepairType("repair_retry_backoff", "60s", (type) -> verify(probe, times(1)).setAutoRepairRetryBackoff(type.name(), "60s"))
+            forEachRepairType("repair_retry_backoff", "60s", (type) -> verify(probe, times(1)).setAutoRepairRetryBackoff(type.name(), "60s")),
+            forEachRepairType("mixed_major_version_repair_enabled", "false", (type) -> verify(probe, times(1)).setMixedMajorVersionRepairEnabled(false))
```
Since this isn't by repair type now, it can probably just be in its own test like `testHistoryDeleteHostsClearBufferInSec`

tolbertam commented:
Looks good, just tried it out and setautorepairconfig works without a repair type now.  Added one note about tweaking a test, after that I think only thing that's missing now is that it's not present in `nodetool getautorepairconfig` output, mind adding that?  I think just need to add something to `AutoRepairService.getAutoRepairConfiguration()` and that should do it.

himanshujindal committed:
Add config to getRepairConfig output

tolbertam commented on src/java/org/apache/cassandra/service/AutoRepairService.java:
```diff
@@ -102,6 +102,7 @@ public String getAutoRepairConfiguration()
         appendConfig(sb, "repair_check_interval", config.getRepairCheckInterval());
         appendConfig(sb, "repair_task_min_duration", config.getRepairTaskMinDuration());
         appendConfig(sb, "history_clear_delete_hosts_buffer_interval", config.getAutoRepairHistoryClearDeleteHostsBufferInterval());
+        appendConfig(sb, "mixed", config.getMixedMajorVersionRepairEnabled());
```
Should have the full name here: `mixed_major_version_repair_enabled`

himanshujindal replied:
I somehow missed that 😭. Fixed.

himanshujindal committed:
Fix typo in config name

tolbertam reviewed (approved):
Looks great! :+1:

jaydeepkumar1984 commented on test/unit/org/apache/cassandra/repair/autorepair/AutoRepairUtilsTest.java:
```diff
@@ -232,6 +236,52 @@ public void testGetHostIdsInCurrentRing_multiple_nodes()
+            String.format("%d.%d", NodeVersion.CURRENT.cassandraVersion.major - 1, 0));
+        ClusterMetadataTestHelper.addEndpoint(2, new NodeVersion(differentCassandraVersion,
+                                                                 NodeVersion.CURRENT_METADATA_VERSION));
+        // With the same major versions, but different minor versions, we should still see this function return true
```
nit typo: With the same minor versions, but different major versions

himanshujindal replied:
Fixed.

himanshujindal committed:
Address feedback
